monitoring/src/monitors/chain_monitor.rs

```diff
@@ -0,0 +1,164 @@
+    // skip serialization to JSON for ws
+    #[serde(skip_serializing)]
+    // timestamp for fisrt block header occurrence
+    start: Instant,
+    // time to download headers and operations for cycle
+    duration: Option<f32>,
+}
+
+// monitoring statistics for bootraping
+#[derive(Debug, Clone, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ChainMonitor {
+    // number of block per cycle
+    blocks_per_cycle: usize,
+    // cycle is used to measure time
+    chain: Vec<Cycle>,
+}
+
+impl ChainMonitor {
+    pub fn new() -> Self {
+        Self {
+            blocks_per_cycle: 4096,
+            chain: vec![],
+        }
+    }
+
+    // get cycle id number
+    fn cycle_id(&mut self, block_level: usize) -> usize {
+        block_level / self.blocks_per_cycle
+    }
+
+    // extend chain to support new cycles
+    fn update_chain(&mut self, block_level: usize) {
+        // cycle number
+        let cycle_id = self.cycle_id(block_level);
+        // get numner of cycles in chain
+        let chain_len = self.chain.len();
+        // create all missing cycles for chain stats
+        if cycle_id >= chain_len {
+            let mut chain_append = (0..(cycle_id - chain_len + 1))
+                .map(|id| -> Cycle {
+                    Cycle {
+                        id: chain_len + id,
+                        headers: 0,
+                        operations: 0,
+                        applications: 0,
+                        start: Instant::now(),
+                        duration: None,
+                    }
+                })
+                .collect();
+            // append created chain stats
+            self.chain.append(&mut chain_append);
+        };
+        // println!("[block] header level: {:.9} cycle_id: {:.4} chain_len: {:.3}",  block_level, cycle_id, chain_len);
+    }
+
+    // process block header
+    pub fn process_block_header(&mut self, block_level: usize) {
+        // extend chain to support new cycles
+        self.update_chain(block_level);
+        // cycle number
+        let cycle_id = self.cycle_id(block_level);
+        // use reference to update cycle stats
+        let cycle: &mut Cycle = &mut self.chain[cycle_id];
+        // increment block headers value
+        cycle.headers += 1;
+    }
+
+    // process block operations
+    pub fn process_block_operations(&mut self, block_level: usize) {
+        // extend chain to support new cycles
+        self.update_chain(block_level);
+        // cycle number
+        let cycle_id = self.cycle_id(block_level);
+        // use reference to update cycle stats
+        let cycle: &mut Cycle = &mut self.chain[cycle_id];
+        // increment block operations value
+        cycle.operations += 1;
+        // save duration for cycle download time at end of cycle
+        if cycle.operations == self.blocks_per_cycle {
+            // save duration in seconds
+            cycle.duration = Some(cycle.start.elapsed().as_secs_f32())
```

Comment:
Instead of using number of seconds you can create `Instant::now()` here. After doing that you will benefit from compiler type checking your time operations.

Reply:
Default `std::time::Duration` serialization is too wild to be used in any standard API.